tsdb/ooo_head.go

```diff
@@ -26,16 +27,28 @@ import (
 // Samples are stored uncompressed to allow easy sorting.
 // Perhaps we can be more efficient later.
 type OOOChunk struct {
-	samples []sample
+	samples    []sample
+	tombstones *tombstones.MemTombstones
+	seriesRef  storage.SeriesRef
 }
 
 func NewOOOChunk() *OOOChunk {
-	return &OOOChunk{samples: make([]sample, 0, 4)}
+	// return &OOOChunk{samples: make([]sample, 0, 4)}
+	return &OOOChunk{
+		samples:    make([]sample, 0, 4),
+		tombstones: tombstones.NewMemTombstones(),
+	}
 }
 
 // Insert inserts the sample such that order is maintained.
-// Returns false if insert was not possible due to the same timestamp already existing.
+// Returns false if insert was not possible due to the same timestamp already existing
+// or the timestamp falls within a tombstone interval.
 func (o *OOOChunk) Insert(t int64, v float64) bool {
+	// Check if this timestamp is in a tombstone interval before inserting.
+	if o.tombstones.HasTimestamp(o.seriesRef, t) {
```

Comment:
No need to do this check. We want very fast writes.

Example case:
``` 
insertOOO(t0,t1,t2,t3,t4,t5)
deleteOOO(t1,t2) // just adds tombstone to ooohead
insertOOO(t1,t2) // will probably complain about duplicates
queryOOO(t0,t5) // returns t0,t3,t4,t5 (missing t1,t2)
``` 

The block querier takes into account tombstones :+1: